### ISSUE TITLE
@:persistent is not always enough persistent

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#7778]
When developing domkit I hit into the following problem:

I have classes that are being `@:build`. For each of these builds I create a `MetaComponent` instance that is stored into a `@:persistent static var Map<String,MetaComponent>`, for later resolution.

Things works quite good since when the classes are modified, `@:build` is triggered and this leads to updating the Map with the new version.

However, when I modify some macro (it doesn't have to be domkit macro, just a macro already loaded as part of the macro context), it will discard the macro context and rebuild it.

At this point, the persistant data has been discarded, and the `@:build` classes will not be built again because they are cached. So even if I do a resolveType, I'm unable to get back a MetaComponent.

I think we should be able to specify that @:persistent will be restored even on macro context rebuild, making it a de-facto "compiler context storage" that we discussed before. 

Another option would be to introduce two primitives :   `Context.getStorageValue(key:String)`, `Context.setStorageValue(key:String,value:Dynamic)` that will be stored as part of the cached context (one cache per set of defines).

**Comment by ncannasse** [comment]
ping @Simn @back2dos

**Comment by Simn** [comment]
I'm still very confused by this. If we rebuild something in the macro context, we do it for a reason, namely that it or one of its dependencies changed. Subverting that would put the context in an invalid state where there could be a stored value from an older compilation mixed into the current compilation. That sounds like a recipe for disaster to me.

I'm also not sure what you mean when you say "discard the macro context" because I don't believe we discard it as a whole anymore. Modifying macro A does not discard macro B unless there's a dependency. You can see that in an example like this:

```haxe
// Macro.hx
class Macro {
	static public macro function test() {
		$type("typing Macro.test");
		trace("Running Macro.test");
		return macro null;
	}
}

```

```
// Main.hx
class Main {
	static public function main() {
		Macro.test();
		test();
	}

	macro static function test() {
		$type("Typing Main.test");
		trace("Running Main.test");
		return macro null;
	}
}
```

If you modify Main.hx, the `$type` in `Macro.hx` is not executed again, meaning it's not typed again.

So the real question here is why your stuff is being re-typed.

**Comment by ncannasse** [comment]
Yes, the macros itself are not retyped, but the macro context is discarded and rebuilt to account for the Main.test change, which means a @:persistent in Macro class will be reset.

Or, I need to be able to store a list (actually map) of the classes that have been built, and there's a mismatch between our compiler cache and the macro context persistant cache for the reason above.

**Comment by back2dos** [comment]
Yeah, this is a minefield :D

Anyway, for this particular case, I think the best solution would be to be able to attach persistent cache *to specific types*. From my experience, the type cache is far more reliable than the macro module cache. I would propose an API along these lines:

```haxe
class Cache {
  public final type:haxe.macro.Type;
  public function get(key:String):Dynamic;
  public function set(key:String, value:Dynamic):Void;

  static public function forType(type:haxe.macro.Type):Cache;
  static public function getAllWithKey(key:String):Array<Cache>;// not sure this is needed, but you could use it to retrieve a list of all domkit components if that's what you want?
}
```

As soon as `type` is reloaded, the cache is gone.

This would also be nice for caching the result of expensive `@:genericBuild` macros. Grab the cache for the parameter and look if the work has been done already. As soon as that type gets reloaded, cache is gone and the macro automatically falls through to rebuilding a new type. The a really nightmare-inducing chunk of tink_macro that I would love to delete in favor of this ;)

**Comment by Simn** [comment]
> Yes, the macros itself are not retyped, but the macro context is discarded and rebuilt to account for the Main.test change, which means a @:persistent in Macro class will be reset.

It's not:


```haxe
// Macro.hx
class Macro {
	@:persistent static var counter = 0;

	static public macro function test() {
		trace("Running Macro.test " + counter++);
		return macro null;
	}
}
```

```haxe
// Main.hx
class Main {
	static public function main() {
		#if !macro
		Macro.test();
		test();
		#end
	}

	macro static function test() {
		trace("Running Main.test");
		return macro null;
	}
}
```

```
>start haxe.exe --wait 6666

>haxe -main Main --interp --connect 6666
Macro.hx:5: Running Macro.test 0
Main.hx:10: Running Main.test

>haxe -main Main --interp --connect 6666

>touch Main.hx

>haxe -main Main --interp --connect 6666
Macro.hx:5: Running Macro.test 1
Main.hx:10: Running Main.test

>haxe -main Main --interp --connect 6666
```

I thought that was the whole idea behind `@:persistent`.

**Comment by back2dos** [comment]
Well, as far as I understand Nicolas, his @:persistent Map get's reset. Which it shouldn't, as far as I understand you.

Beyond that, being able to persist something *for as long as a given type is not invalidated* would be very helpful. If you consider Nicolas's case, this set up leaks memory (although probably not enough to matter), because it will keep data even for components that you remove from the build.

**Comment by ncannasse** [comment]
Maybe I didn't understood correctly @:persistent, but it definitely gets
discarded when I modify the macro that have @:persistent or one of the
macro which rely on it :)
It's then not as a big problem as I thought it would be. But I think the
per-type cache Juraj suggested would make sense.

**Comment by ncannasse** [comment]
OTOH if we solve #7448 this will make sure that the cache gets entirely
rebuilt

**Comment by Simn** [comment]
It's unclear to me if there's anything I have to do here for 4.0. I like the idea of tying invalidation to a type, which I think comes down to allowing to set up a module dependency by hand.

**Comment by back2dos** [comment]
To me it's still not clear if `@:persistent` works as advertised or not. And if doesn't (as Nicolas would seem to indicate), then something should be done about it for 4.0.

**Comment by Simn** [comment]
I'm happy to discuss this, but I'll need a concrete example that shows what isn't working as expected.

**Comment by ncannasse** [comment]
My problem has been solved with #7448
